pkg/plugin/validator.go

```diff
@@ -0,0 +1,189 @@
+	}
+	v := Validator{
+		cfg:        cfg,
+		decodedPem: pk,
+	}
+	for _, opt := range opts {
+		opt(&v)
+	}
+	if v.client == nil {
+		v.client = github.NewClient(nil)
+	}
+	if v.githubApp == nil {
+		ghCfg := githubapp.NewConfig(v.cfg.GitHubAppID, v.cfg.GitHubAppInstallationID, v.decodedPem, githubapp.WithJWTTokenCaching(1*time.Minute))
+		v.githubApp = githubapp.New(ghCfg)
+	}
+	return &v, nil
+}
+
+// MatchIssue parses issue info from provided issueURL and validate if the issue is valid.
+func (v *Validator) MatchIssue(ctx context.Context, issueURL string, opts ...githubapp.ConfigOption) error {
+	info, err := parseIssueInfoFromURL(issueURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse issueURL: %w", err)
+	}
+
+	t, err := v.getAccessToken(ctx, info.RepoName)
+	if err != nil {
+		return fmt.Errorf("failed to get access token: %w", err)
+	}
+	v.client = v.client.WithAuthToken(t)
+
+	return v.validateIssue(ctx, info)
+}
+
+// validateIssue verifies if the issue exists and the issue is open.
+func (v *Validator) validateIssue(ctx context.Context, pi *pluginGithubIssue) error {
+	issue, err := v.getGithubIssue(ctx, pi)
+	if err != nil {
+		return fmt.Errorf("failed to get issue info: %w", err)
+	}
+	if s := issue.GetState(); s != "open" {
+		return fmt.Errorf("issue is in state: %s, please make sure to use an open issue", s)
+	}
+	return nil
+}
+
+// getGithubIssue gets the provided issue's info from github api.
+func (v *Validator) getGithubIssue(ctx context.Context, pi *pluginGithubIssue) (*github.Issue, error) {
+	issue, _, err := v.client.Issues.Get(ctx, pi.Owner, pi.RepoName, pi.IssueNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get issue: %w", err)
+	}
+	return issue, nil
+}
+
+// getAccessToken gets an access token with issue read permission to the repo
+// which contains the issue.
+func (v *Validator) getAccessToken(ctx context.Context, repoName string) (string, error) {
+	tr := githubapp.TokenRequest{
+		Repositories: []string{repoName},
+		Permissions: map[string]string{
+			"issues": "read",
+		},
+	}
+	var tokenResp ExchangeResponse
+	resp, err := v.githubApp.AccessToken(ctx, &tr)
+	if err != nil {
+		return "", fmt.Errorf("failed to get access token: %w", err)
+	}
+
+	if err := json.Unmarshal([]byte(resp), &tokenResp); err != nil {
+		return "", fmt.Errorf("error unmarshal resp: %w", err)
+	}
+	return tokenResp.AccessToken, nil
+}
+
+// parseGithubIssue parses issue info from Issue URL.
+func parseIssueInfoFromURL(issueURL string) (*pluginGithubIssue, error) {
+	if match, _ := regexp.MatchString(issueURLPatternRegExp, issueURL); !match {
+		return nil, fmt.Errorf("invalid issue url, issueURL doesn't match pattern: %s", issueURLPatternRegExp)
+	}
+	u, err := url.Parse(issueURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse provided issue url: %w", err)
+	}
+	arr := strings.Split(u.Path, "/")
+
+	issueNumber, err := strconv.Atoi(arr[4])
```

Comment:
Bug? Need to check len first?

Reply:
if the url passed the regexp check, it's garenteed that arr[4] exists.

Reply:
I see. Let's add a comment to signal boost it. In that case, could you also capture the owner/repo, etc. directly from the regex matching? @drevell is the regex expert

Reply:
comment added, @drevell is OOO, created issue https://github.com/abcxyz/jvs-plugin-github/issues/4 to track this.